engine/client/client.go

```diff
@@ -325,112 +327,171 @@ func (c *Client) startEngine(ctx context.Context) (rerr error) {
+		Method: http.MethodGet,
+		URL: &url.URL{
+			Scheme: "http",
+			Host:   "dagger",
+			Path:   engine.SessionAttachablesEndpoint,
+		},
+		Header: headers,
+		Host:   "dagger",
+	}
+	if err := req.Write(conn); err != nil {
+		return nil, fmt.Errorf("write request: %w", err)
+	}
 
-	// registry auth
-	bkSession.Allow(authprovider.NewDockerAuthProvider(config.LoadDefaultConfigFile(os.Stderr), nil))
+	resp, err := http.ReadResponse(bufio.NewReader(conn), req)
+	if err != nil {
+		return nil, fmt.Errorf("read response: %w", err)
+	}
+	if resp.Body != nil {
+		defer resp.Body.Close()
+	}
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		var respBody []byte
+		if resp.Body != nil {
+			respBody, _ = io.ReadAll(resp.Body)
+		}
+		return nil, fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(respBody))
+	}
 
-	// host=>container networking
-	bkSession.Allow(session.NewTunnelListenerAttachable(ctx, nil))
+	// We tell the server that we have fully read the response and will now switch to serving gRPC
+	// by sending a single byte ack. This prevents the server from starting to send gRPC client
+	// traffic while we are still reading the previous HTTP response.
+	if _, err := conn.Write([]byte{0}); err != nil {
+		return nil, fmt.Errorf("write ack: %w", err)
+	}
 
-	// connect to the server, registering our session attachables and starting the server if not
-	// already started
-	ctx, cancel := context.WithCancelCause(ctx)
-	c.eg.Go(func() error {
-		err := bkSession.Run(c.internalCtx, func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
-			c, err := grpchijack.Dialer(c.bkClient.ControlClient())(ctx, proto, engine.ClientMetadata{
-				RegisterClient:            true,
-				ClientID:                  c.ID,
-				ClientVersion:             engine.Version,
-				ServerID:                  c.ServerID,
-				ClientSecretToken:         c.SecretToken,
-				ClientHostname:            hostname,
-				UpstreamCacheImportConfig: c.upstreamCacheImportOptions,
-				UpstreamCacheExportConfig: c.upstreamCacheExportOptions,
-				Labels:                    c.labels,
-				CloudToken:                os.Getenv("DAGGER_CLOUD_TOKEN"),
-				DoNotTrack:                analytics.DoNotTrack(),
-			}.AppendToMD(meta))
-			return c, err
-		})
-		cancel(errSessionTerminated)
-		return err
-	})
+	return sessionSrv, nil
+}
 
-	// Try connecting to the session server to make sure it's running
-	c.httpClient = &http.Client{Transport: &http.Transport{
-		DialContext: c.DialContext,
-		// connection re-use in combination with the underlying grpc stream makes
-		// managing the lifetime of connections very confusing, so disabling for now
-		// TODO: For performance, it would be better to figure out a way to re-enable this
-		DisableKeepAlives: true,
-	}}
-
-	// there are fast retries server-side so we can start out with a large interval here
-	if err := retry(ctx, 3*time.Second, func(elapsed time.Duration, ctx context.Context) error {
```

Comment:
Nice seeing this retry logic going away. Guessing it's just not needed anymore, for reasons similar to the other comment?

Reply:
Yep, once we've successfully made any request from a given client in a given session, we know that the client+session state are fully initialized. So because we have already made a request successfully at this point (to hook up the attachables) we know we're good to go and don't need this "retry a request" stuff anymore.